### PR TITLE
PLT-6965 jira integration (plus plugin scaffolding)

### PR DESCRIPTION
**Pull request by ccbrown**
#### Summary

This pull request adds a JIRA integration.

**At a feature-level...** the only real criteria was multi-project support. [csduarte/mattermost-jira](https://github.com/csduarte/mattermost-jira) supports multiple projects and I've been told is a project we tend to direct people to. So the goal for the initial commit was just feature parity, and the behavior is identical except for message formatting, which is slightly different (better).

**At a technical level...** I was told the long-term goal was for this to be a server plugin as described by the [developer toolkit roadmap](https://docs.mattermost.com/developer/toolkit.html). So there was technical criteria of "written in Go and easily separable from the rest of the codebase". Thus the integration is written as a compile-time plugin of sorts. The JIRA-specifics are isolated and abstracted away from the Mattermost application and vice-versa. There's a narrow and well-defined boundary between the two, which can serve as scaffolding for a more permanent plugin architecture where the boundary can be crossed by dynamically linked binaries, RPC calls, etc.

**From a user's perspective...** to get up and running, you need to edit your config file:

```
"PluginSettings": {
    "Plugins": {
        "jira": {
            "Secret": "supersecretstring",
            "UserId": "rd49ehbqyjytddasoownkuqrxe"
        }
    }
}
```

You need to define some secret, which protects the webhook endpoint and a user, which will be making the posts on JIRA's behalf.

Once you've done that, you just need to add a webhook to JIRA that looks like this:

> https://**theserver**/plugins/jira/webhook?secret=**supersecretstring**&team=**theteam**&channel=**thechannel**

So the process is pretty much the same as for [csduarte/mattermost-jira](https://github.com/csduarte/mattermost-jira), except instead of creating an incoming webhook and deploying a server, you create a user and edit your config file.

**From my perspective...** I don't think deploying your typical bot as a server plugin will be advantageous over HTTP-based bots for a very long time. It's gonna be pretty challenging to get the development / installation / upgrade process for server plugins low-friction enough that they're developed on more than a need-to-use basis.

On the other hand, I think it would be a really good goal to create a plugin architecture capable of fully supporting our enterprise features as plugins. That's the kind of use-case where this sort of thing will really shine.

#### Ticket Link
https://mattermost.atlassian.net/browse/PLT-6965

#### Checklist
- [ ] Added or updated unit tests (required for all new features)

### CODE REVIEW

File: app/channel.go [hunk 1]
```diff
@@ -1198,3 +1198,39 @@ func GetPinnedPosts(channelId string) (*model.PostList, *model.AppError) {
 		return result.Data.(*model.PostList), nil
 	}
 }
+
+func GetOrCreateChannel(teamId, userId, channelNameOrId string) (*model.Channel, *model.AppError) {
```
Comment (ccbrown) [inline app/channel.go]: no functional changes here. more or less cut / pasted from _app/webhook.go_ so i could re-use it

File: app/plugin/jira/plugin.go [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+		panic(err)
+	}
+	api.Router().HandleFunc("/webhook", p.handleWebhook).Methods("POST")
+}
+
+func (p *Plugin) handleWebhook(w http.ResponseWriter, r *http.Request) {
+	if p.Configuration.Secret == "" || p.Configuration.UserId == "" {
+		http.Error(w, "This plugin is not configured.", http.StatusForbidden)
+	} else if subtle.ConstantTimeCompare([]byte(r.URL.Query().Get("secret")), []byte(p.Configuration.Secret)) != 1 {
+		http.Error(w, "You must provide the configured secret.", http.StatusForbidden)
```
Comment (ccbrown) [inline app/plugin/jira/plugin.go]: not localized. these strings aren't user-facing, and i didn't want to add jira-related strings to our main i18n files. i'd prefer to wait until we figure out a good strategy that allows plugin-specific translations to be fully encapsulated by the plugins

File: app/plugin/jira/plugin_test.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+	"net/http"
+	"net/http/httptest"
+	"os"
+	"testing"
+
+	"github.com/mattermost/platform/app/plugin/plugintest"
+	"github.com/mattermost/platform/model"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
```
Comment (ccbrown) [inline app/plugin/jira/plugin_test.go]: making a pitch for testify here. these three subpages make tests so much easier to write and output so much more useful.

File: glide.lock [hunk 4]
```diff
@@ -166,6 +166,12 @@ imports:
   version: e57e3eeb33f795204c1ca35f56c44f83227c6e66
 - name: github.com/spf13/viper
   version: c1de95864d73a5465492829d7cb2dd422b19ac96
+- name: github.com/stretchr/testify
+  version: 69483b4bd14f5845b5a1e55bca19e954e827f1d0
```
Comment (ccbrown) [inline glide.lock]: before merging, if testify looks alright, i'll need to `glide install`. left them out of the review to reduce pr noise

File: app/plugin/jira/webhook.go [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+			"#### " +
+			"[![{{.User.Name}}]({{index .User.AvatarUrls \"16x16\"}})]({{.JIRAURL}}/secure/ViewProfile.jspa?name={{urlquery .User.Name}}) " +
+			"[{{.User.DisplayName}}]({{.JIRAURL}}/secure/ViewProfile.jspa?name={{urlquery .User.Name}})** " +
+			"{{.Verb}} " +
+			"[![{{.Issue.Fields.IssueType.Name}}]({{.Issue.Fields.IssueType.IconURL}})]({{.JIRAURL}}/browse/{{.Issue.Key}}) " +
+			"[[{{.Issue.Key}}] {{.Issue.Fields.Summary}}]({{.JIRAURL}}/browse/{{.Issue.Key}})**" +
+			"{{range .ChangeLog.Items}}" +
+			"\n* **{{title .Field}}**: {{if eq .Field \"description\"}}\n  ```\n{{indent .ToString \"  \"}}\n  ```{{else}}~~{{.FromString}}~~ {{.ToString}}{{end}}" +
+			"{{end}}" +
+			"{{if .Comment.Body}}\n\n> {{.Comment.Body}}{{end}}",
```
Comment (ccbrown) [inline app/plugin/jira/webhook.go]: screen shot 2017-07-12 at 2 59 16 am

**Comment by ccbrown** [comment]
test failure looks like someone else's bug

**Comment by ccbrown** [comment]
added a system console page:

screen shot 2017-07-14 at 2 18 39 am

@jasonblais, i think it's slightly different from what you suggested in the interest of keeping code and organization a little more plugin-like: instead of adding a "plugins" page within the "integrations" section, i added a "jira" page within a "plugins" section. it can be changed within minutes if there's actually opposition to that though. 1/5

**Comment by coreyhulen** [comment]
I wonder if we should add `Enable`/`Disable` in the config so you can temporarily disable JIRA without deleting your keys.  It also makes it obvious that it's enabled.

**Comment by coreyhulen** [comment]
How does a user find the TeamID?  should it be `team name or team id` like channel? 

We have the view info channel model to find the channel id.  I wonder if we need something similar.

**Comment by coreyhulen** [comment]
I guess the same is true for UserId.

**Comment by jasonblais** [comment]
@ccbrown Can you help look at the build failures when you get the chance? I'll set up a test instance and give this one a try.

I agree with Corey though: I'm not actually sure how I would retrieve the team ID or user ID. I thought the user ID would be on the Users page but we only seem to display the username. For comparison, most of our CLI commands support team and username for that reason https://docs.mattermost.com/administration/command-line-tools.html

As for the System Console page, the change is okay. I'll evaluate the design once I've tested the plugin from start to finish and have a better idea on how it works.

**Comment by jasonblais** [comment]
As a note though...this is really awesome work and I'm quite excited about this feature. Great work getting this so quickly.

**Comment by ccbrown** [comment]
Adding an enable / disable would be easy enough.

As for the user and team ids, I think the nicest thing would be to add a nice user picker and a URL generator. This sort of thing:

screen shot 2017-07-14 at 5 09 42 pm

but for users and teams. That might be a bit too much work for this review though. So if I can't easily do that, I'll see about allowing team or user names since I don't know of a user-friendly way to get the ids either.

Fixed my eslint errors, so hopefully the build will work now.

**Comment by jasonblais** [comment]
Some initial thoughts:

1 - It makes sense to add the Enable/Disable toggle. It's more consistent with our other System Console pages, and makes it more obvious if the plugin is active or not.
2 -  As discussed above, propose we switch it so either 
 - team url
 - username
 - channel url
works. We can leave support for team_id/channel_id/user_id for now, but in the UI we'll probably refer to the username/team_url/channel_url only, since those are more easily accessible.

I actually got stuck setting up the Jira plugin as the only way to retrieve the team ID is via the DB (I think).

As for your suggestion of using a channel picker -- that's something we can consider but that would be V2. We'd have to consider how that works with multiple teams and channels, and I worry it could get a bit complicated if an instance has multiple Jira hooks running across teams & channels.

Having a single webhook URL they can use for each of the Jira hooks is sufficient for V1.

----

Some other questions: 

3 - Should the secret be hidden with ************* once the settings are saved? Similar to our other secrets in the System Console?
4 - Why do we have the user_id config setting? Why not include it as part of the URL similar to team and channel IDs (or names once we switch it over)? That way admins can overwrite it easily if they want.
5 - Is the reason why we're tying the plugin to a config setting (instead of doing it in the Integrations page) that you don't have to generate a separate secret for each team?

I have a few other questions that I'll bring up next week, mostly related to permissions. But I'll think through them while I'm working on related stuff for Joram's API token.

**Comment by ccbrown** [comment]
@jasonblais 

I've been doing some polish here. Most of it was done before your comment, but I think it happens to address a lot of the concerns:

screen shot 2017-07-16 at 1 39 50 am

1. Done.
2. Switched. I had/have a slight preference for supporting user / team ids only, but those definitely require more UX work (e.g. a URL builder) than I really want to do for a first commit. I did also went ahead and add user autocomplete though since that's relatively easy.
3. It can't really be. It has to be made accessible. Otherwise there'd be no way for the admin to craft the URLs to plug into JIRA. Maybe a more complex implementation could do something a little more advanced that keeps the secrets a little more secret though.
4. I'm not sure there's a use-case for that, and I think configuring it once is easier in most cases (especially now with autocomplete). Another big reason is that I think in the future (hopefully relatively short-term), the user shouldn't be a configuration parameter at all. Posts should have the appearance of being posted from the plugin itself rather than requiring a user to be made and customized beforehand. This is more similar to how Slack apps work and is overall much faster for users to get started with.
5. I think we were just trying to limit the scope and do something simple. When I first put in this PR, there were no plans to build any UI at all. :P This should absolutely become a team integration, but I think we could do it much more elegantly and with much less tech debt if we tackled some things on the developer toolkit roadmap beforehand:

    * *The ability to build webapp client plugins to override existing UI components [...], modify/extend client drivers to interact with custom server API endpoints, and add whole new UI views in predetermined places.*
    * *The ability to build Go 1.8 server plugins to [...] have some form of database access*
    * *A system or architecture to combine the above plugins into one easy-to-share and easy-to-install package.*

**Comment by jasonblais** [comment]
1/2 - thanks! Although I'm not convinced on the user picker. We can discuss it.
3 - ah yes, that's true.. And I guess it's not really all that different from the webhook URL we display for incoming webhooks. Which everyone can see (as supposed to just System Admins seeing the Jira webhook URL)
4 - okay that's true. It'd be like a Jira bot posting to channels in the future. We can keep the current design for now and then consider removing the setting once we have support for such bot accounts.
5 - makes sense :)

-----

Unfortunately I couldn't get this to work :/

Attached are the System Console and Jira settings. Did I do something wrong?

image

image

Jira Secret: `vepx0bNjbBo69i+Gn3TQXuBy7V28oPJN`
Jira URL: `http://i-0faf5423006c659da.spinmint.com/plugins/jira/webhook?secret=vepx0bNjbBo69i%2BGn3TQXuBy7V28oPJN&team=pr6918&channel=town-square`

**Comment by ccbrown** [comment]
@jasonblais 

Yeah, but you can blame whoever owns these Spinmint servers. None of them boot up with the right site URL configured (They're all missing the port number.). 😕 

Try plugging this URL into JIRA:

```
http://i-0faf5423006c659da.spinmint.com:8065/plugins/jira/webhook?secret=vepx0bNjbBo69i%2BGn3TQXuBy7V28oPJN&team=pr6918&channel=town-square
```

**Comment by jasonblais** [comment]
@ccbrown I tried that too actually but Jira gave me this error.

image

So I guess it can't really be tested on a spinmint then..

**Comment by ccbrown** [comment]
TODO:
* Make empty strings in change logs not render as "~~~~".
* Remove DM support for now to avoid complications related to [this JIRA bug](https://jira.atlassian.com/browse/JRASERVER-42579)?

**Comment by ccbrown** [comment]
@csduarte 

1. For plugins in general, I'm not sure we can solve this problem at the plugin API level. This will probably be up to plugin implementations, since a variety of mechanisms will be used for plugin-to-external-service authentication.

    For this plugin, this is already easy if you don't care about a few minutes of plugin downtime while you update your JIRA webhooks. But if you need zero plugin downtime, I think the way to solve this is to make it a team integration and allow it to store multiple configurations. Then teams can have transition periods of overlapping secrets.

2. Rate-limiting is a good concern for the future of plugins in general, and can probably be addressed at the plugin API level. We could definitely enforce configurable rules such as "plugins can create no more than 60 posts per channel, per minute". Did you have any specific types of rate-limiting in mind?

3. Disabling the plugin would short-circuits the requests so they just respond with an error code immediately. We'll definitely have the ability to install / enable / disable / uninstall plugins easily. Is there something beyond this that would be helpful in these scenarios?

While this plugin is currently being compiled directly into the server, an upcoming goal is to move it back out of the server, into a self-contained package that can just be dropped onto servers. My personally vision is for Go plugins that run in a separate process, as an unprivileged user, communicating via RPC (maybe something similar to [hashicorp/go-plugin](https://github.com/hashicorp/go-plugin). That gives us a lot of isolation and the ability to really police what plugins can and can't do, mitigating a lot of the inherited risks involved.

**Comment by coreyhulen** [comment]
@jasonblais @ccbrown I would like to try and push this in for 4.1 and label it Beta/Preview.  Thoughts?

Msg it so people understand the config props may change.

**Comment by ccbrown** [comment]
@coreyhulen 

I'd love to get this in for 4.1 as well, but I'm 0/5 on how we label it.

**Comment by jasonblais** [comment]
Yes, I was going to propose a "Beta" tag as well. And definitely v4.1, I don't think we should delay it.

@ccbrown We can just add it to the System Console section title (instead of `JIRA`, call it `JIRA (Beta)`)

@ccbrown Also, I'm finishing up the last reviews, I'll post these by Monday. Aim is to push it to dev review by Tuesday or Wednesday, then merge by the feature cut-off date which is 7/31

**Comment by jasonblais** [comment]
@ccbrown Okay have fully tested this and prepared proposed UI designs for the System Console page and post formatting.

Proposed changes to current implementation. Most of these we already discussed so there shouldn't really be any surprises, except maybe for 8.

1) Add a value to diagnostics.go on whether a server has the JIRA plugin enabled or not.
2) Add a client-side diagnostic that triggers an event whenever the Jira plugin is used (i.e. post created by the plugin).
3) Remove the “@username” support for sending Jira webhooks to Mattermost DMs. (i.e. don’t escape the URL)
4) Don’t use “reserved” characters in the Jira secret (`!*'();:@+$,/?#`)
5) Make empty strings in change logs not render as "~~~~".
6) Add the “Bot” tag for JIRA webhook posts
 - Make sure it overrides props stuff so it won’t appear in user self searches, etc. 
 - Also make sure that DB queries ran to see what posts were made from a hook/bot using these props will include the JIRA posts.
7) Updates to post formatting and System Console page designs: https://docs.google.com/presentation/d/1ENZMkvnpzoh3mBIuWhKqxUPQx0_TMuLfrYele_o5Yf4/edit#slide=id.g23f8784c1a_0_36 
 - I will ping people to review it on Monday, so the designs are subject to change.
 - Feedback welcome.
8) Remove support for `issue_updated`, or reduce which issue updates trigger posts
 - Right now even the tiniest change (moving a ticket to another sprint, or fixing a typo) triggers a post. We’ll likely get complaints that there are too many updates and that the JIRA hook just spams channels with endless updates.
 - Plus it’s difficult get consistent formatting that works for the different types of updates.
 - If it’s low mana, I’m open to supporting new comments, updates to ticket assignee, and/or updates to ticket status (e.g. open → resolved). Those are useful which are also supported by Slack.
 - But even then, I’m leaning towards just cutting and collecting feedback for V2. Most likely people want this to be configurable (some might want to see comments, but others not, for instance).

**Comment by jasonblais** [comment]
@ccbrown That's a good point and actually agree. 

So for V1, we'd have
 1. `issue_created`
 2. `issue_deleted` if it was previously open
 3. `issue_updated` if closed or re-opened.

That seems reasonable. And admins can configure which of the above three events to trigger from the JIRA webhook page.

**Comment by ccbrown** [comment]
* [x] Add a value to diagnostics.go on whether a server has the JIRA plugin enabled or not.
* [x] Remove the “@username” support for sending Jira webhooks to Mattermost DMs. (i.e. don’t escape the URL)
* [x] Add the “Bot” tag for JIRA webhook posts
* [x] Don’t use “reserved” characters in the Jira secret (!*'();:@+$,/?#)

* [ ] Add a client-side diagnostic that triggers an event whenever the Jira plugin is used (i.e. post created by the plugin).
* [ ] Make empty strings in change logs not render as "~~~~".
* [ ] Updates to post formatting and System Console page designs
* [ ] Remove support for issue_updated, or reduce which issue updates trigger posts

**Comment by jasonblais** [comment]
@ccbrown 

Here are the designs: https://docs.google.com/presentation/d/1bz2gIqYDl0mjCz-pcx8zz3pC5EAvwQTPFuIjNQLQ-ho/edit#slide=id.g24094c60d8_0_0

If you have any questions, let me know.  Let's prioritize work towards this PR so we can merge it by Monday's feature complete day.

Notes:
 1) I have also listed the user auto-complete fixes on the slide
 2) For `Remove support for issue_updated, or reduce which issue updates trigger posts` --> let's keep support for when an issue is reopened or closed, as you suggested. But we can remove others.

**Comment by ccbrown** [comment]
@jasonblais 

I believe I've made all the changes exactly as requested except for two things:

1. I didn't change the header text to "JIRA Plugin (Beta)". I just changed it to "JIRA (Beta)". It seems to me that we shouldn't mention plugins in the front-end yet, giving people the false impression that a plugin system actually exists. But I'll change if desired.

2. I think I need clarification for the client-side diagnostic. My best guess is that you're wanting user clients to send an event any time they view a post made by the plugin for the first time. So one event per every user, per post made. But that sounds really weird to me. And isn't trivial to implement. It seems like a server-side diagnostic would be simpler and possibly more valuable. So can you clarify on exactly what kind of diagnostic event you're wanting?

File: config/config.json [hunk 6]
```diff
@@ -46,10 +46,18 @@
         "TimeBetweenUserTypingUpdatesMilliseconds": 5000,
         "EnablePostSearch": true,
         "EnableUserTypingMessages": true,
-        "EnableChannelViewedMessages": true,
         "EnableUserStatuses": true,
+        "EnableChannelViewedMessages": true,
```
Comment (jasonblais) [inline config/config.json]: Looks like the settings were removed from config.json. Was this intentional? 

Also, there are a few changes to config.json in your diff which might have been unintentional.

File: webapp/plugins/jira/components/settings.jsx [hunk 7]
```diff
@@ -0,0 +1,251 @@
...
+    }
+
+    renderTitle() {
+        return Utils.localizeMessage('admin.plugins.jira', 'JIRA (Beta)');
+    }
+
+    renderSettings() {
+        var webhookDocsLink = (
+            <a
+                href='https://developer.atlassian.com/jiradev/jira-apis/webhooks#Webhooks-configure'
```
Comment (jasonblais) [inline webapp/plugins/jira/components/settings.jsx]: Propose we instead link to https://about.mattermost.com/default-jira-plugin

We can redirect the page either to our own docs or to https://developer.atlassian.com/jiradev/jira-apis/webhooks#Webhooks-configure.

That way if the doc page we redirect to breaks (e.g. the page was moved), all we have to fix is the redirect without changes to the code.

**Comment by jasonblais** [comment]
Thanks @ccbrown! This looks excellent, only a few minor notes below:

1 - My answer to your note: I'm okay with that and makes sense. This is really a JIRA webhook integration, not a plugin per say.

Can you help with one small change by **moving the JIRA section below "Custom Integrations", above "WebRTC (Beta)"**, since it's more related to webhooks than WebRTC and External Services sections.

2 - Re: client-side diagnostic. All I'm really looking for is an event that gets sent when the JIRA webhook makes a post to a channel. I don't care if and when someone has viewed the post, just that it was created.

This is useful given we'll be able to track 
 - how many servers have enabled the JIRA integration
 - how often it's used (i.e. how often a post is created using that integration)

There are two existing events, `api_posts_create` and `api_integrations_used`, which might be related. The first one sends an event whenever a post is created, the second whenever an integration is used.

3 - I couldn't test the integration because spinmint uses a non-standard port. Can you help set up a proxy as last time? I just want to test the following cases:
 - issue created with long description (>4K characters or lots of line breaks)
 - issue created with no user assigned
 - issue updated and deleted + smoke test on other `issue_updated` events not firing a post

4 - Added two comments in-line, see above.

5 - Messaged Asaad about the dropdown UX: There's a couple of minor issues, and I thought to ask his help first so you can focus on the other changes.

PS: Bonus points for having the mattermost URL automatically set in the webhook URL 👍

**Comment by ccbrown** [comment]
@jasonblais 

1. Updated.
2. I'm gonna have to get someone to help with this. The `api_posts_create` and `api_integrations_used` events are sent by the clients whenever the user creates a post or executes a slash command. These and all of the other events we track are all things that are initiated by the user client-side. But these posts aren't initiated by users, and we don't currently have any server-side events that are tracked in this way. I _think_ I might be able to make something work, but I'm really not in a position to confidently implement and test this properly today.
3. Sure: `https://32zanxm6u6.execute-api.us-east-1.amazonaws.com/dev/proxy?url=http%3A%2F%2Fi-0cbfca5a2fa262dbc.spinmint.com%3A8065%2Fplugins%2Fjira%2Fwebhook%3Fsecret%3D_QclLvh6Hn2H6Tgar134B4fLn5Tz71LK%26team%3Dpr6918%26channel%3Dtown-square`. If we have to bring up a new instance or regenerate the secret, just add the URL-encoded (https://www.urlencoder.org/) webhook URL to `https://32zanxm6u6.execute-api.us-east-1.amazonaws.com/dev/proxy?url=`.
4. Updated.

**Comment by jasonblais** [comment]
1. 👍 
2. Ticket here https://mattermost.atlassian.net/browse/PLT-7260
3. See below.
4. 👍 
5. AM submitted updates: https://github.com/mattermost/platform/pull/6918/commits/b515c69fb22a9794d942f140388fb5c399959349



Regarding 3),
 - I noticed that issue descriptions with >4K characters fail to post. Wondering if there's anything we can do about it? One option might be to truncate the `text` field to 4K characters and post with that. We can later figure out a more elegant way of handling long descriptions.
 - Would it be easy to truncate message attachment text to 300 characters before `Show More...` is shown?

**Comment by ccbrown** [comment]
@jasonblais 

Regarding 3, your first bullet is trivial, but I won't be able to update this review until late tonight (Maybe someone else could do it sooner if we'd like?). Your second bullet is also trivial as long as we make it apply to all message attachments. Making that reduced limit only apply to these message attachments and not others would probably warrant a separate ticket.

**Comment by jasonblais** [comment]
> Your second bullet is also trivial as long as we make it apply to all message attachments.

Yes, exactly. It would apply to all message attachments.

@ccbrown I'm also fine creating a separate ticket to fix these last two edge cases before Friday's code complete, and merging this PR as-is. They're not critical bugs that would prevent merging the PR for today's feature complete.

Let me know if you'd prefer that. If so, can you help merge conflicts and I can ask Joram and Corey to help with dev reviews.

**Comment by ccbrown** [comment]
@jasonblais Yeah, let's make a separate ticket so we can get the dev reviews started as soon as possible on this one.

**Comment by jasonblais** [comment]
Done: https://mattermost.atlassian.net/browse/PLT-7263

Will request dev review next

File: app/channel.go [hunk 8]
```diff
@@ -1198,3 +1198,19 @@ func GetPinnedPosts(channelId string) (*model.PostList, *model.AppError) {
 		return result.Data.(*model.PostList), nil
 	}
 }
+
+func GetDirectChannel(userId1, userId2 string) (*model.Channel, *model.AppError) {
```
Comment (jwilander) [inline app/channel.go]: No action required, functionally it's the same as `CreateDirectChannel()` but it does the DB read first which is probably less strain on the database. Might be better to combine them in the future but doesn't have to block this PR

File: app/plugin/api.go [hunk 9]
```diff
@@ -0,0 +1,34 @@
+package plugin
```
Comment (jwilander) [inline app/plugin/api.go]: Legal requirement, all new files need to start with:
```
// Copyright (c) 2017-present Mattermost, Inc. All Rights Reserved.
// See License.txt for license information.
```

File: Makefile [hunk 10]
```diff
@@ -539,38 +539,21 @@ nuke: clean clean-docker
 setup-mac:
 	echo $$(boot2docker ip 2> /dev/null) dockerhost | sudo tee -a /etc/hosts
 
+NON_ENTERPRISE_PACKAGES=./api/... ./api4/... ./app/... ./cmd/... ./einterfaces/... ./imports/... ./jobs/... \
```
Comment (coreyhulen) [inline Makefile]: This an un-related change?  next time lets submit it as a separate PR.

File: app/plugin/jira/plugin_test.go (hunk 3 above)
Comment (coreyhulen) [inline app/plugin/jira/plugin_test.go]: Next time lets queue and item for the developer meeting to discuss.

File: app/plugin/api.go (hunk 9 above)
Comment (ccbrown) [inline app/plugin/api.go]: There should be a test for this: https://mattermost.atlassian.net/browse/PLT-7264

File: Makefile (hunk 10 above)
Comment (ccbrown) [inline Makefile]: Hmm... I thought I had responded, but I guess I never submitted it.

This change was not unrelated. This had to be changed in some way to get vet running against the new package. I very much disagree that this should have been a separate PR and think there might be discussions to be had about opportunistic refactoring.
